proposals/extensions_v2.md

```diff
@@ -0,0 +1,401 @@
+
+```c#
+extension E
+{
+    // Instance method form, replaces `public static int M<X>(this SomeType<X> val, ...) { } 
+    public int M<X>(...) for SomeType<X> val { }
+
+    // Property form:
+    // Auto-properties, or usage of 'field' not allowed as extensions do not have instance state.
+    public int Count<X> for SomeType<X> val { get { ... } }
+
+    // Event form:
+    // Note: would have to be the add/remove form.
+    // field-backed events would not be possible as extensions do not have instance state.
+    public event Action E<X> for SomeType<X> val { add { } remove { } }
+
+    // Indexer form:
+    public int this<T>[int index] for SomeType<X> val { get { ... } }
+
+    // Operator form:
+    // note: no SomeType<X> val, an operator is static, so it is not passed an instance value.
+    public static SomeTypeX<X> operator+ <X>(SomeType<X> s1, SomeType<X> s2) for SomeType<X> { ... }
+    
+    // Conversion form:
+    // note: no SomeType<X> val, an operator is static, so it is not passed an instance value.
+    public static implicit operator SomeTypeX<X>(int i) for SomeType<X> { ... }
+
+    // Constructor form:
+    // note: no SomeType<X> val, an operator is static, so it is not passed an instance value.
+    public SomeType<X>() for SomeType<X> { }
+
+    // *Static* extension method (not possible today).  Called as `Assert.Boo("", "")`
+    // note: no `Assert val`, a static method is not passed an instance value.
+    public static bool Boo(string s1, string s2) for Assert { }
+
+    // Static extensions properties, indexers and events are all conceptually supportable.
+    // Though we can decide which are sensible to have or not.
+    // Static extensions can having backing static fields. Static extension properties can use `field`.
+    
+    // Nested types must be supported to maintain compatibility with existing static classes with extension members in them.
+}
+```
+
+Without specifying the full grammar changes, the intuition is that we are making the following changes:
+
+```g4
+for-clause
+    | 'for' parameter
+    ;
+
+parameter (no change)
+    | attributes? modifiers? type identifier ...
+    ;
+
+extension
+    | attributes? modifiers? 'extension' identifier { member_declaration* }
+    ;
+
+// For method/property/indexer/operator/constructor/event declarations
+// we are augmenting its syntax to allow type-parameters
+// (if not already allowed) and a for-clause. For example:
+property-declaration
+    | attributes? modifiers identifier type-parameters for-clause property-body
+    ;
+    
+compilation-unit-member
+    | ...
+    | extension
+    ;
+ 
+namespace-declaration-member
+    | ...
+    | extension
+    ;
+```
+
+The use of `parameter` means all of the following are legal, with the same semantics that that classic extension methods have today:
+
+```c#
+for ref Span<T> span
```

Comment:
Personal taste, I'd put parens on anything more complicated than just-a-type, similarly to lambda parameters. Parens in `for (ref readonly Span<T> span)` help visual parsing, but they don't seem needed for `for Span<T>`.

Reply:
Another example to illustrate the value of parens: `for [Attribute1][Attribute2] Type name`